lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+        try check_errno(res);
+        const fd = @intCast(i32, res);
+        assert(fd >= 0);
+        errdefer os.close(fd);
+
+        // Kernel versions 5.4 and up use only one mmap() for the submission and completion queues.
+        // This is not an optional feature for us... if the kernel does it, we have to do it.
+        // The thinking on this by the kernel developers was that both the submission and the
+        // completion queue rings have sizes just over a power of two, but the submission queue ring
+        // is significantly smaller with u32 slots. By bundling both in a single mmap, the kernel
+        // gets the submission queue ring for free.
+        // See https://patchwork.kernel.org/patch/11115257 for the kernel patch.
+        // We do not support the double mmap() done before 5.4, because we want to keep the
+        // init/deinit mmap paths simple and because io_uring has had many bug fixes even since 5.4.
+        if ((p.*.features & linux.IORING_FEAT_SINGLE_MMAP) == 0) {
+            return error.IO_UringKernelNotSupported;
+        }
+
+        // Check that the kernel has actually set params and that "impossible is nothing".
+        assert(p.*.sq_entries != 0);
+        assert(p.*.cq_entries != 0);
+        assert(p.*.cq_entries >= p.*.sq_entries);
+
+        // From here on, we only need to read from params, so pass `p` by value for convenience.
+        // The completion queue shares the mmap with the submission queue, so pass `sq` there too.
+        var sq = try SubmissionQueue.init(fd, p.*);
+        errdefer sq.deinit();
+        var cq = try CompletionQueue.init(fd, p.*, sq);
+        errdefer cq.deinit();
+
+        // Check that our starting state is as we expect.
+        assert(sq.head.* == 0);
+        assert(sq.tail.* == 0);
+        assert(sq.mask.* == p.*.sq_entries - 1);
+        // Allow flags.* to be non-zero, since the kernel may set IORING_SQ_NEED_WAKEUP at any time.
+        assert(sq.dropped.* == 0);
+        assert(sq.array.len == p.*.sq_entries);
+        assert(sq.sqes.len == p.*.sq_entries);
+        assert(sq.sqe_head == 0);
+        assert(sq.sqe_tail == 0);
+
+        assert(cq.head.* == 0);
+        assert(cq.tail.* == 0);
+        assert(cq.mask.* == p.*.cq_entries - 1);
+        assert(cq.overflow.* == 0);
+        assert(cq.cqes.len == p.*.cq_entries);
+
+        // Alles in Ordnung!
+        return IO_Uring {
+            .fd = fd,
+            .sq = sq,
+            .cq = cq,
+            .flags = p.*.flags
+        };
+    }
+
+    pub fn deinit(self: *IO_Uring) void {
+        assert(self.fd >= 0);
+        // The mmaps depend on the fd, so the order of these calls is important:
+        self.cq.deinit();
+        self.sq.deinit();
+        os.close(self.fd);
+        self.fd = -1;
+    }
+
+    /// Returns a vacant SQE, or an error if the submission queue is full.
+    /// We follow the implementation (and atomics) of liburing's `io_uring_get_sqe()` exactly.
```

Comment:
There are many schemes for giving out SQEs, I'm not sure this is the best one for zig's standard library: I'm curious if there is any reusable pieces out of e.g. the general purpose allocator.

Reply:
Absolutely, and people can always use their own helpers for the scheme they prefer, but this is also a defacto liburing method, so I wanted it here as a starting point. The `sqe_tail` and flushing logic is also quite critical to amortizing the atomic stores on some architectures, and has a nice safety property, i.e. you can run the ring with SQPOLL, and get a pointer to an sqe, and modify it, without updating the SQ tail and without racing the kernel. If the name bothers you, we could call this `reserve_free_sqe()`... and then later also add a `free_reserved_sqes()` helper if someone wants to throw the queued batch away.

Reply:
Also, just to be clear for anyone else following, this does not allocate anything, only returning a pointer to a free SQE and bumping our shadow `sqe_tail` pointer.